src/rgw/rgw_lua.cc

```diff
@@ -137,5 +143,146 @@ int delete_script(rgw::sal::RGWRadosStore* store, const std::string& tenant, opt
 
   return 0;
 }
+
+#ifdef WITH_RADOSGW_LUA_PACKAGES
+
+const std::string PACKAGE_LIST_OBJECT_NAME = "lua_package_allowlist";
+
+namespace bp = boost::process;
+
+int add_package(rgw::sal::RGWRadosStore* store, optional_yield y, const std::string& package_name, bool allow_compilation) {
+  // verify that luarocks can load this oackage
+  const auto p = bp::search_path("luarocks");
+  if (p.empty()) {
+    return -ECHILD;
+  }
+  bp::ipstream is;
+  const auto cmd = p.string() + " search --porcelain" + (allow_compilation ? " " : " --binary ") + package_name;
+  bp::child c(cmd,
+      bp::std_in.close(),
+      bp::std_err > bp::null,
+      bp::std_out > is);
+
+  std::string line;
+  bool package_found = false;
+  // TODO: yield on reading the output
+  while (c.running() && std::getline(is, line) && !line.empty()) {
+    package_found = true;
+  }
+  c.wait();
+  auto ret = c.exit_code();
+  if (ret) {
+    return -ret;
+  }
+
+  if (!package_found) {
+    return -EINVAL;
+  }
+  
+  // add package to list
+  const bufferlist empty_bl;
+  std::map<std::string, bufferlist> new_package{{package_name, empty_bl}};
+  librados::ObjectWriteOperation op;
+  op.omap_set(new_package);
+  ret = rgw_rados_operate(*(store->getRados()->get_lc_pool_ctx()), 
+      PACKAGE_LIST_OBJECT_NAME, &op, y);
+
+  if (ret < 0) {
+    return ret;
+  } 
+  return 0;
+}
+
+int remove_package(rgw::sal::RGWRadosStore* store, optional_yield y, const std::string& package_name) {
+  librados::ObjectWriteOperation op;
+  op.omap_rm_keys(std::set<std::string>({package_name}));
+  const auto ret = rgw_rados_operate(*(store->getRados()->get_lc_pool_ctx()), 
+    PACKAGE_LIST_OBJECT_NAME, &op, y);
+
+  if (ret < 0) {
+    return ret;
+  }
+
+  return 0;
+}
+
+int list_packages(rgw::sal::RGWRadosStore* store, optional_yield y, packages_t& packages) {
+  constexpr auto max_chunk = 1024U;
+  std::string start_after;
+  bool more = true;
+  int rval;
+  while (more) {
+    librados::ObjectReadOperation op;
+    packages_t packages_chunk;
+    op.omap_get_keys2(start_after, max_chunk, &packages_chunk, &more, &rval);
+    const auto ret = rgw_rados_operate(*(store->getRados()->get_lc_pool_ctx()),
+      PACKAGE_LIST_OBJECT_NAME, &op, nullptr, y);
+  
+    if (ret < 0) {
+      return ret;
+    }
+
+    packages.merge(packages_chunk);
+  }
+ 
+  return 0;
+}
+
+int install_packages(rgw::sal::RGWRadosStore* store, optional_yield y, packages_t& failed_packages, std::string& output) {
```

Comment:
generalizing for all of (add, remove, list);  packages appear to be known via omap, but are present in a local path;  can you explain how this works when we have a group of rgws on different hosts?

Reply:
each RGW install the packages on its own host, later on to be used by the lua VM.

now I see that we have a problem with multiple RGWs running on the same host. will have to create separate directory for each RGW so they won't step on one another.

Reply:
you can use radosgw's `EntityName` as a stable/unique filesystem path. `conf->name.to_str()` will give you something like `client.rgw.0`